### PR TITLE
[EA] People directory

### PR DESCRIPTION
**Pull request by oetherington**
This PR adds the beta version of the EA people directory ([Figma](https://www.figma.com/file/IRUVQ5D2mKxoW0xZnhXIWv/2024-Q1?node-id=1463%3A12795&mode=dev)).

After deploying, we need to run the following (which I've already done on dev but not staging or prod):
```
./scripts/serverShellCommand.sh "Globals.regenerateUserProfileUpdatedAt()"
./scripts/serverShellCommand.sh "Globals.elasticConfigureIndex('Users')"
./scripts/serverShellCommand.sh "Globals.elasticExportCollection('Users')"
```
The first step takes 20-25 minutes, but the second two are much quicker.

Screenshot 2024-04-09 at 14 34 55

Screenshot 2024-04-09 at 14 35 15

Screenshot 2024-04-09 at 14 36 06

Screenshot 2024-04-09 at 14 36 23

Screenshot 2024-04-09 at 14 36 31

Screenshot 2024-04-09 at 14 36 44

Screenshot 2024-04-09 at 14 36 52

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206918564163649) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by agnesstenlund** [comment]
probably didn't state this anywhere in the designs but I was imagining "OR" filter for this dropdown (think it's currently "AND"?)
Screenshot 2024-03-25 at 18 53 07

**Comment by agnesstenlund** [comment]
also, this is looking really fecking amazing 🥇

**Comment by agnesstenlund** [comment]
these two we can make slimmer by default since they're never above a certain width :) that would leave more space for other columns

Screenshot 2024-03-25 at 19 09 01

**Comment by agnesstenlund** [comment]
Meta-comment: I'd try to make some of the dropdowns more general components, to make it easier to reuse them in places like these later on (so not making them "people-directory-dropdown" but instead something like "multi-select-dropdown")
Screenshot 2024-03-26 at 11 03 08
Screenshot 2024-03-26 at 11 03 59

Screenshot 2024-03-26 at 11 13 01
Screenshot 2024-03-26 at 11 13 22
Screenshot 2024-03-26 at 11 13 11

**Comment by agnesstenlund** [comment]
> Do you think all the filters should be "OR"?

I do, always "OR" inside of one filter. But "AND" between different filters. So "Location: Stockholm OR Oslo" AND "Org is CEA or 80k" would show people in Stockholm or Oslo who also work at either CEA or 80k

**Comment by oetherington** [comment]
@agnesstenlund Another open question: I'm currently excluding the current user from the results, but I'm not sure if this is correct or not

**Comment by agnesstenlund** [comment]
> I'm currently excluding the current user from the results, but I'm not sure if this is correct or not

I wouldn't! I think it's good knowing you're showing up + when you show up. So everybody sees the same thing

**Comment by agnesstenlund** [comment]
We should also have some smart loading setup, like showing more rows as you scroll, or let people click a "Load more" at the bottom (slightly less nice, I think)

**Comment by agnesstenlund** [comment]
Really like the hovers <3

**Comment by agnesstenlund** [comment]
maybe this is a local environment thing, but I don't get any results for the locations that show up in the list–I assume we only show locations in the list that people have added? 😮 
Screenshot 2024-04-04 at 14 07 46
Screenshot 2024-04-04 at 14 07 32

**Comment by agnesstenlund** [comment]
just had another look, it's starting to feel _so_ great 🥲 

Some things:

- [ ] For the beta version, I think it'd be really cool to change the "NEW" label to "BETA", so people feel special + more inclined to write feedback :)
Screenshot 2024-04-05 at 14 31 40

- [ ] We should also add a "Give feedback" button to the beta version, either link to Intercom or to a Google form I write
Screenshot 2024-04-05 at 14 50 54

- [ ] I like this width a lot for the smaller columns, but realising now they need a bit more padding in between each other
Screenshot 2024-04-05 at 14 19 04

- [ ] Would be nice to show a hover text when hovering the message icon–would do "Send message" which will hopefully make it more clear it's a message and not an email
Screenshot 2024-04-05 at 14 21 39

- [ ] Change to "Profile last updated" as per request by Toby. Lets keep "Profile updated" in the table though so the column doesn't get too wide
Screenshot 2024-04-05 at 14 23 13

- [ ] **Not needed for beta** but there are probably some scenarios where we do want the slim columns to act a bit flexible again, like in this circumstance: 
Screenshot 2024-04-05 at 14 27 57

- [ ] **Also not needed for beta** but these don't have a different hover + focused state
Screenshot 2024-04-05 at 14 29 01
Screenshot 2024-04-05 at 14 29 08

**Comment by agnesstenlund** [comment]
would be cool to add the number of search results like this for now:
Screenshot 2024-04-08 at 14 56 22

**Comment by agnesstenlund** [comment]
- [ ] oh also the little caret is the wrong color when there is something in the filter:

Screenshot 2024-04-08 at 14 37 31

**Comment by agnesstenlund** [comment]
- [ ] wouldn't say this is needed for beta, but if it's easy we could implement the empty state
image

**Comment by agnesstenlund** [comment]
- [ ] we also need a different hover state for filled out filters 🙏 

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/305d8a5a-8070-4765-8374-c0b8537e64de

**Comment by agnesstenlund** [comment]
the recent changes are so good, thank youu 🙌

**Comment by agnesstenlund** [comment]
- [ ] btw "Clear search" in empty state should also clear the filters

**Comment by agnesstenlund** [comment]
- [ ] one more tiny thing, I think it'd be slightly nicer if we only show the number of results after you've added any filters or search words
Screenshot 2024-04-09 at 12 40 19

**Comment by agnesstenlund** [comment]
- [ ] Before we ship this to beta we should probably hide it on mobile, alternatively show the sidebar thing but then on the actual directory say "This is a beta feature and only works on desktop" instead of showing the directory itself

**Comment by agnesstenlund** [comment]
Other than that I think this is ready to ship to beta. We could add the topic interests before, but that needs a bit of a redesign in settings so would probably exclude that for the v1 beta

**Comment by agnesstenlund** [comment]
+ link the "share your thoughts" to this google form: https://forms.gle/3RpvUhiwyKFzMSju9 

is there anything else left to do @oetherington?

**Comment by oetherington** [comment]
Great! Thanks @agnesstenlund!

I'll make those last few changes then open this up for code review (which might take a while as it's a very big PR). There's nothing else I can think of right now.

**Comment by agnesstenlund** [comment]
also the skeleton loading is looking soo nice ✨

Any chance each column could keep their width? Not a big deal if not, it still looks really nice. But that'd avoid a tiny layout shift:

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/f61c5798-561a-4a96-8f7b-05c9f2cf8c9e

**Comment by oetherington** [comment]
@agnesstenlund I'll have a think about it, but it's probably not trivial (definitely doable though).

File: packages/lesswrong/components/peopleDirectory/PeopleDirectoryFilters.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+          title="Columns"
+          icon="ViewColumns"
+          style="button"
+          className={classes.columnsList}
+        >
+          {columns.filter(({hideable}) => hideable).map((state) => (
+            <PeopleDirectorySelectOption state={state} key={state.value} />
+          ))}
+        </PeopleDirectoryFilterDropdown>
+      </div>
```
Comment (Will-Howard) [inline packages/lesswrong/components/peopleDirectory/PeopleDirectoryFilters.tsx:75]: I would like a "reset to default" option on this one (cc @agnesstenlund )

**Review by Will-Howard — COMMENTED** [review]
Generally looks good, I think there are a few blocking things, I'll do a second pass and note which comments are blocking. Here are few bugs which don't fit anywhere specific:
- When I navigate away and back to the page (i.e. don't do a full reload) it then doesn't autoload more items on scroll, so it's stuck on the first few it loaded
- I got this error when I tried searching "ollie etheri..." (edit: I also got it when filtering by location, the final screenshot is from that case):
Screenshot 2024-04-15 at 11 48 13
Screenshot 2024-04-15 at 11 48 24
Screenshot 2024-04-15 at 12 24 28
- It would be really nice to have scroll restoration when you navigate away and back, but I know this would be quite a big change to do. Failing that, maybe making it easy to open people's profiles in a new tab to nudge people to do that as the default (cc @agnesstenlund)?
- Not sure if I said this already but given the proliferation of acronym org names, I think making this keep the case it was originally in would be good (not sure where the casing is being applied)
Screenshot 2024-04-15 at 11 54 27

Blocking things:
- [ ] Navigating away and back to the page causes it to not load more items on scroll
- [ ] The "Cannot read properties of undefined" bug when searching on dev (it may turn out this is a dev only issue)

File: packages/lesswrong/components/peopleDirectory/PeopleDirectoryHeading.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  },
+});
+
+export const PeopleDirectoryHeading = ({column, classes}: {
+  column: PeopleDirectoryColumn,
+  classes: ClassesType<typeof styles>,
+}) => {
+  const {sorting, setSorting} = usePeopleDirectory();
+
+  const onToggleSort = useCallback(() => {
```
[the remark this answers is not in the document]
Comment (agnesstenlund) [reply packages/lesswrong/components/peopleDirectory/PeopleDirectoryHeading.tsx:29]: Agree, thanks for raising this. We're considering a "most filled out profiles + high karma"-thing as "default sort". Do you think seeing it sorted by "Default" would be enough or would that leave you wondering what that means?

File: packages/lesswrong/components/peopleDirectory/PeopleDirectoryUserCell.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    alignItems: "center",
+    gap: "8px",
+    width: "100%",
+  },
+  name: {
+    ...textCellStyles(theme),
+    flexGrow: 1,
+  },
+  message: {
+    display: "none", // See `PeopleDirectoryUserCell` for logic to show this
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/components/peopleDirectory/PeopleDirectoryUserCell.tsx:19]: Yes, but the relevant css `:hover` selector is in the other file

File: packages/lesswrong/components/peopleDirectory/PeopleDirectorySelectOption.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  });
+
+  const onClick = useCallback(() => {
+    onToggle();
+    captureEvent();
+  }, [onToggle, captureEvent]);
+
+  const {ForumIcon} = Components;
+  return (
+    <div onClick={onClick} className={classNames(classes.root, className)}>
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/components/peopleDirectory/PeopleDirectorySelectOption.tsx:59]: At the moment it's just the "check" icon from Heroicons. We could add some padding? This is how it looks with 1px of padding:
Screenshot 2024-04-23 at 12 23 37

File: packages/lesswrong/components/hooks/useSearchableMultiSelect.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      return hits ?? [];
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error("Facet search error:", e);
+      captureException(e);
+      return [];
+    }
+  }, [captureSearch, facetField]);
+
+  const getWithCache = useLRUCache<string, Promise<string[]>>(fetchSuggestions);
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/components/hooks/useSearchableMultiSelect.ts:95]: Hmm - it's just matching the behaviour of the main searchbox. I think I'm just going to leave it for now and think about it more for when we leave beta.

**Comment by oetherington** [comment]
@agnesstenlund I think all comments are now addressed. I also gated it so that only beta users will see the people directory in the main menu - I think this was the plan for releasing?

Still outstanding questions from above:

1. Do we want a "reset to default" option for the columns dropdown?
2. Do we need some indication of which sorting is applied?
3. Does it matter that there's currently no user map? Options:
  a. Revert removing the map from the groups page
  b. Add the map to the people directory be release
  c. Do nothing now, add the map to the people directory soon after release
4. Do we want to adjust the styles of the checkboxes in the filter dropdowns?

**Comment by agnesstenlund** [comment]
Thank you!! I think none of these are blocking shipping to beta, but I think most of them should be done soon after.

> 1. Do we want a "reset to default" option for the columns dropdown?

that seems good, although not necessary for beta

> 2. Do we need some indication of which sorting is applied?

I think we do, although I'm fine shipping to beta without

> 3. Does it matter that there's currently no user map?

If either reverting or adding is really quick, I'd do that (there is a small risk people would send feedback about it missing). I do think so few people visit the groups page that it's also not a big deal to have it gone for a couple of weeks.


> 4. Do we want to adjust the styles of the checkboxes in the filter dropdowns?

yes, but IMO also not blocking releasing to beta

something that would be nice to have (to increase discoverability for the beta) is adding this to search:
image

but also wouldn't consider that blocking shipping to beta ☝️

**Comment by agnesstenlund** [comment]
And it's correct only beta users should see it in the main menu 👍

**Comment by oetherington** [comment]
@agnesstenlund Great! I just added the "Reset" button for the columns dropdown, reverted the change that removes the community map so it's back on the groups page, and added the link to the main users search page. I think we're ready to go now?

(Also CC @Will-Howard for code approval)

**Comment by agnesstenlund** [comment]
something weird happened to the LHS 😮  the people directory only shows for a sec when you refresh

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/46988dfc-2ab2-4d7a-969a-5d344b7a8c4f

**Comment by agnesstenlund** [comment]
Looks great to me! 🚢

**Comment by oetherington** [comment]
That makes sense that it was fixed after a `yarn install` - we just did a react upgrade (sorry, I should have warned you but totally forgot!)

File: packages/lesswrong/components/peopleDirectory/PeopleDirectoryHeading.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/components/peopleDirectory/PeopleDirectoryHeading.tsx:29]: I think having there be a default sorting that is somewhat opaque is ok (with or without a "Default" label), it's when there is a sorting applied that I think the arrow on the relevant column should show as up or down

**Review by Will-Howard — APPROVED** [review]
LGTM. If you're not in a huge rush to deploy it I think showing up and down arrows when sorting by a certain column is the most important thing to improve.

Also is there a reason not to allow sorting by karma? I would assume people would want that, also maybe by "Profile updated" but that seems less important

**Comment by agnesstenlund** [comment]
I quite like this UI pattern (that highlights the column title + the direction it's getting sorting)
Screenshot 2024-04-25 at 17 43 00

or just showing the arrow in one direction in the primary text color:
Screenshot 2024-04-25 at 17 45 47

I also agree it probably make sense to be able to sort by both karma and profile updated, probably looks too opinionated not being able to (I had some idea around how you should only be able to filter by karma, not sort, to not make it too easy to see who's got the most or least, but I think I probably endorse transparency more)

**Pull request merged**